parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/HadoopStreams.java

```diff
@@ -0,0 +1,116 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.parquet.hadoop.util;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.parquet.Log;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.apache.parquet.io.SeekableInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.nio.ByteBuffer;
+
+public class HadoopStreams {
+
+  private static final Log LOG = Log.getLog(SeekableInputStream.class);
+
+  private static final Class<?> byteBufferReadableClass;
+
+  private static final Constructor<SeekableInputStream> h2SeekableConstructor;
+
+  static {
+    byteBufferReadableClass = getReadableClass();
+    h2SeekableConstructor = getH2SeekableConstructor();
+  }
+
+  public static SeekableInputStream wrap(FSDataInputStream stream) {
+    if (byteBufferReadableClass != null && h2SeekableConstructor != null &&
+        byteBufferReadableClass.isInstance(stream) &&
+        supportsByteBufferReads(stream)) {
+      try {
+        return h2SeekableConstructor.newInstance(stream);
+      } catch (InstantiationException e) {
+        LOG.warn("Could not instantiate H2SeekableInputStream, falling back to byte array reads", e);
+        return new H1SeekableInputStream(stream);
+      } catch (IllegalAccessException e) {
+        LOG.warn("Could not instantiate H2SeekableInputStream, falling back to byte array reads", e);
+        return new H1SeekableInputStream(stream);
+      } catch (InvocationTargetException e) {
+        throw new ParquetDecodingException(
+            "Could not instantiate H2SeekableInputStream", e.getTargetException());
+      }
+    } else {
+      return new H1SeekableInputStream(stream);
+    }
+  }
+
+  private static final ByteBuffer ZERO_LEN_BYTE_BUFFER = ByteBuffer.wrap(new byte[0]);
+
+  private static boolean supportsByteBufferReads(FSDataInputStream stream) {
+    // FSDataInputStream implements ByteBufferReadable, but may throw
+    // UnsupportedOperationException if it isn't actually supported. This tests
+    // whether the method throws by trying to read a 0-length buffer, which has
+    // no effect on the stream when it is supported.
+    try {
+      stream.read(ZERO_LEN_BYTE_BUFFER);
+      return true;
+    } catch (UnsupportedOperationException e) {
+      return false;
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  private static Class<?> getReadableClass() {
+    try {
+      return Class.forName("org.apache.hadoop.fs.ByteBufferReadable");
+    } catch (ClassNotFoundException e) {
+      return null;
+    } catch (NoClassDefFoundError e) {
+      return null;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Class<SeekableInputStream> getH2SeekableClass() {
+    try {
+      return (Class<SeekableInputStream>) Class.forName(
+          "org.apache.parquet.hadoop.util.H2SeekableInputStream");
```

Comment:
We don't need to expose the class, so I don't think we need to expose its name anywhere. That pretty much removes the need to separate this out as a constant.